# CAO Points Analysis
***

In [1]:
# import required libraries
import requests as rq             # Convenient HTTP requests.
import re                         # Regular expressions.
import datetime as dt   
import pandas as pd
import numpy as np

# Get the current date and time in desired string format
nowstr = dt.datetime.now().strftime('%Y%m%d_%H%M%S')

<br>

**Error on server (reference function below)** <br>
The Server indicated one should decode as per: Content-Type: text/html; charset=iso-8859-1. However, one line of the html uses<br> 
\x96 which isn't defined in iso-8859-1. Therefore the similar decoding standard cp1252 will be  used. This standard is very similar but<br>
includes \x96.

In [2]:

def website_data_handler(url, regex):
    
    response = rq.get(url)                                  # Make a get request to 2021 CAO points URL
    response.encoding = 'cp1252'                            # Specify encoding method as cp1252 despir what server recommended
    
    no_lines = 0                                            # Keep track of courses counted in loop
    
    with open('data/cao2021.csv', 'w') as f:                    # Open the csv file for writing         
        for line in response.iter_lines():                          # Loop through lines of the response 
            dline = line.decode('cp1252')                           # decode the line 
            if regex.fullmatch(dline):                          # Match only the lines representing courses.       
                no_lines = no_lines + 1                                 # Add one to the lines counter if match occured      
                dline_adj = re.sub(" ", "  ", dline, count = 1)         # Substitute first space character with double space     
                linesplit = re.split('   +', dline_adj)                 # Split the line on three or more spaces.  
                f.write(','.join(linesplit) + '\n')                     # Rejoin the substrings with commas in between.         
    print(f"Total number of lines is {no_lines}.")         # Print the total number of processed lines
    

In [3]:
url_2021 = 'http://www2.cao.ie/points/l8.php'                                                       # Make a get request to 2021 CAO points URL
regex_course_2021 = re.compile(r'([A-Z]{2}[0-9]{3})  (.*)(#?)(?:([AQ]{3})|([V]{1})|([0-9]{3}))(\*?) *')     # define regular expression
website_data_handler(url_2021, regex_course_2021)                                                       # pass paremeters into parse & write function


# Regular Expression breakdown:                Explanation for what characters will be parsed:    
#**************************************        ***************************************************************************************************
# ([A-Z]{2}[0-9]{3})                           first 2 chars (A-Z), next 3 chars, 0-9 followed by two spaces
# (.*)                                         Matches any single character except newline zero or more times
# (?:([AQ]{3})|([V]{1})|([0-9]{3}))            Match (3 x [AQ]) or (1 x V) or (3 x [0-9]) zero or one time
# (\*?) *'                                    \(escaped) "*" so dont treat it as quantifier, treat it by itself. Zero or one * character "?" 
# space *                                     any number of spaces



Total number of lines is 923.


In [4]:
#new_col_order = ['Year', 'CATEGORY (i.e.ISCED description)', 'HEI', 'COURSE TITLE',              
                 #'COURSE CODE2','R1 POINTS', 'R2 POINTS', 'EOS Mid-point', 'EOS', 
                 #'R1 Random *',  'R2 Random*', 'EOS Random *', 'Test/Interview #'] 


def data_formatter_2021():
    
    # read csv from folder and pass list of column headers
    read_path = 'data/cao2021.csv'
    header_list_2021 = ['COURSE CODE', 'COURSE TITLE', 'R1 POINTS', 'R2 POINTS']       
    df = pd.read_csv(read_path, names = header_list_2021)                                                                                                                                               
    
    df.insert(0, 'YEAR', 2021) 
    
    # insert 3 new conditional columns via string operations 
    # explanation : (if col contains string x, return x, else return y)
    df['R1 Random*'] = np.where(df['R1 POINTS'].str.contains('\*'), '*', '')              
    df['R2 Random*'] = np.where(df['R2 POINTS'].str.contains('NaN'), '',               
                         np.where(df['R2 POINTS'].str.contains('\*'), '*', ''))
    
    df['Test/Interview #'] = np.where((df['R1 POINTS'].str.contains('#')) | 
                                      (df['R2 POINTS'].str.contains('#')), '#', '')  
    
    #df_cols = df.columns.tolist()
    
    cols = ['a', 'b', 'c']
    num = [0, 0, 0]
    df.assign(**dict(zip(cols, num)))
    

    
    # remove '*' and '#' from points columns  
    df['R1 POINTS'].replace(regex = ['\*', '#'], value = '', inplace = True)            
    df['R2 POINTS'].replace(regex = ['\*', '#'], value = '', inplace = True)
    return df
    
    
df_2021 = data_formatter_2021()
df_2021.head(50)

YEAR COURSE CODE                                       COURSE TITLE  \
0   2021       AL801     Software Design for Virtual Reality and Gaming   
1   2021       AL802  Software Design in Artificial Intelligence for...   
2   2021       AL803  Software Design for Mobile Apps and Connected ...   
3   2021       AL805    Computer Engineering for Network Infrastructure   
4   2021       AL810                                 Quantity Surveying   
5   2021       AL820                 Mechanical and Polymer Engineering   
6   2021       AL830                                    General Nursing   
7   2021       AL832                              Mental Health Nursing   
8   2021       AL835                                       Pharmacology   
9   2021       AL836                       Nutrition and Health Science   
10  2021       AL837            Sports Science with Exercise Physiology   
11  2021       AL838                                      Biotechnology   
12  2021       AL839                                       Microbiology   
13  2021       AL840                            Pharmaceutical Sciences   
14  2021       AL841                Athletic and Rehabilitation Therapy   
15  2021       AL842                              Bioveterinary Science   
16  2021       AL843               Physical Activity and Health Science   
17  2021       AL849                          Culinary Entrepreneurship   
18  2021       AL850                                           Business   
19  2021       AL851                                   Business and Law   
20  2021       AL852                                         Accounting   
21  2021       AL854    Sport Management (with international placement)   
22  2021       AL855  Hospitality Management (with international pla...   
23  2021       AL857                                  Digital Marketing   
24  2021       AL859                             International Business   
25  2021       AL860                               Social Care Practice   
26  2021       AL861             Animation and Illustration (portfolio)   
27  2021       AL863             Graphic and Digital Design (portfolio)   
28  2021       AL864                     Early Years Care and Education   
29  2021       AL870                                 Applied Psychology   
30  2021       CW018                                 Business (options)   
31  2021       CW028                 Early Childhood Education and Care   
32  2021       CW038                                    Art (portfolio)   
33  2021       CW058                       Tourism and Event Management   
34  2021       CW068  Applied Social Studies in Professional Social ...   
35  2021       CW078       Sustainable Farm Management and Agribusiness   
36  2021       CW088                   Visual Communications and Design   
37  2021       CW108                                  Science (options)   
38  2021       CW128                             Brewing and Distilling   
39  2021       CW138                         Sport and Exercise Science   
40  2021       CW148                          Strength and Conditioning   
41  2021       CW188          Sport Rehabilitation and Athletic Therapy   
42  2021       CW208                         Computer Games Development   
43  2021       CW238                               Software Development   
44  2021       CW248                  Information Technology Management   
45  2021       CW258                         Cybercrime and IT Security   
46  2021       CW268    Computing in Interactive Digital Art and Design   
47  2021       CW438                             Construction (options)   
48  2021       CW468                           Architectural Technology   
49  2021       CW478                                  Civil Engineering   

   R1 POINTS R2 POINTS R1 Random* R2 Random* Test/Interview #  
0        300       NaN                                         
1        313       NaN                            

# 2020 data

In [5]:
def data_handler_2020():
    
    # read in excel 2020 data 
    read_url = 'http://www2.cao.ie/points/CAOPointsCharts2020.xlsx'
    
    
    df = pd.read_excel(read_url, skiprows = 10)     
    #df = pd.read_excel('data/CAOPointsCharts2020.xlsx', skiprows = 10)  

    
    df.insert(0, 'YEAR', 2020) 
    df = df.loc[df.loc[:, 'LEVEL'] == 8]                                                   # re-define df with level 8 courses filtered   
    df.drop(['avp', 'v', 'LEVEL'], axis = 1, inplace = True)                               # drop 'avp' and 'v' columns as these are not in 2019/2021 data
    df = df.iloc[:,:-8]                                                                    # drop last 8 empty cols dropped 
       
    
    # define col order list
    new_col_order = ['YEAR', 'CATEGORY (i.e.ISCED description)', 'HEI', 'COURSE TITLE',              
                     'COURSE CODE2','R1 POINTS', 'R2 POINTS', 'EOS Mid-point', 'EOS', 
                     'R1 Random *',  'R2 Random*', 'EOS Random *', 'Test/Interview #']  
    # re-order columns
    df = df.reindex(columns = new_col_order)                                                
                       
    # re-name columns from dict key to dict value 
    df.rename(columns = {'CATEGORY (i.e.ISCED description)': 'CATEGORY', 'Hei': 'HEI', 
                        'COURSE CODE2': 'COURSE CODE', 'EOS': 'FINAL SEASON POINTS',
                         'EOS Mid-point': 'MID SEASON POINTS', 'EOS Random *': 
                         'FINAL SEASON POINTS RANDOM*'}, inplace = True)      

    
    col_str_replace = ['R1 POINTS', 'R2 POINTS', 'MID SEASON POINTS', 'FINAL SEASON POINTS']              # define list cols for string replace operation 
    df[col_str_replace] = np.where(df[col_str_replace] == '#+matric', 'NaN', df[col_str_replace])         # replace '#+matric' with 'NaN'      
   
    return df
    


data_2020 = data_handler_2020()
data_2020

YEAR                                           CATEGORY  \
0     2020                        Business and administration   
1     2020                      Humanities (except languages)   
2     2020                                               Arts   
3     2020                                               Arts   
4     2020                                               Arts   
...    ...                                                ...   
1455  2020                                               Arts   
1460  2020  Information and Communication Technologies (ICTs)   
1461  2020  Information and Communication Technologies (ICTs)   
1462  2020                                  Personal services   
1463  2020                 Engineering and engineering trades   

                                     HEI  \
0                       American College   
1                       American College   
2     National College of Art and Design   
3     National College of Art and Design   
4     National College of Art and Design   
...                                  ...   
1455   Waterford Institute of Technology   
1460   Waterford Institute of Technology   
1461   Waterford Institute of Technology   
1462   Waterford Institute of Technology   
1463   Waterford Institute of Technology   

                                           COURSE TITLE COURSE CODE R1 POINTS  \
0                                International Business       AC120       209   
1                                          Liberal Arts       AC137       252   
2      First Year Art & Design (Common Entry,portfolio)       AD101       NaN   
3     Graphic Design and Moving Image Design (portfo...       AD102       NaN   
4     Textile & Surface Design and Jewellery & Objec...       AD103       NaN   
...                                                 ...         ...       ...   
1455                                     Arts (options)       WD200       AQA   
1460                       Software Systems Development       WD210       279   
1461                                 Creative Computing       WD211       271   
1462                    Recreation and Sport Management       WD212       270   
1463           Mechanical and Manufacturing Engineering       WD230       253   

     R2 POINTS MID SEASON POINTS FINAL SEASON POINTS R1 Random * R2 Random*  \
0          NaN               280                 209         NaN        NaN   
1          NaN               270                 252         NaN        NaN   
2          NaN               NaN                 NaN         NaN        NaN   
3          NaN               NaN                 NaN         NaN        NaN   
4          NaN               NaN                 NaN         NaN        NaN   
...        ...               ...                 ...         ...        ...   
1455       AQA               336                 AQA         NaN        NaN   
1460       NaN               337                 279         NaN        NaN   
1461       NaN               318                 271         NaN        NaN   
1462       NaN               349                 270         NaN        NaN   
1463       NaN               369                 253         NaN        NaN   

     FINAL SEASON POINTS RANDOM* Test/Interview #  
0                            NaN              NaN  
1                            NaN              NaN  
2                            NaN                #  
3                            NaN                #  
4                            NaN                #  
...                          ...              ...  
1455                         NaN              NaN  
1460                         NaN              NaN  
1461                         NaN              NaN  
1462                         NaN              NaN  
1463                         NaN              NaN  

[1027 rows x 13 columns]

# 2019 data

In [26]:
# save raw pdf to disc in binary mode (i.e: makes no changes to file)
raw_pdf_2019 = 'data/cao_2019_raw_data.pdf'
with open(raw_pdf_2019, 'wb') as f:  
    f.write(resp.content)

In [32]:
import urllib3
import pdfplumber
import io


def scrape_pdf_from_website(url):
    http = urllib3.PoolManager()                                   # instansiate pool manager object for sending requests
    temp_binary = io.BytesIO()                                     # create instance for storing binary data
    temp_binary.write(http.request("GET", url).data)               # request url data and write to temp_binary storage
    
    data_final = ''  
    with pdfplumber.open(temp_binary) as pdf:                      # open storage object      
        for page in pdf.pages:                                         # iterate pages of pdf object
            data = page.extract_text()                                 # extract text of page
            data_final += data                                         # concatenates text to data_final variable
    return data_final




# request url for raw pdf
url  = 'http://www2.cao.ie/points/lvl8_19.pdf'
pdf_text = scrape_pdf_from_website(url)

In [50]:
def data_handler_2019(regex, line_splitter, text):
    
    df_data = []                                                   # placeholder to add list of rows to

    for line in text:                                              # iterate file lines  
        for match in re.finditer(regex, line):                         # iterate matches each line      
            course_code = match.group()[0:6].strip()                               # isolate course code
            rest_of_line =  match.group()[6:].strip()                               # isolate all of line after course code
            rest_of_line = re.split(line_split_conditions ,rest_of_line)            # split lines using passed criteria
            
    df = pd.DataFrame(dataForDataFrame)                            # convert array to dataframe
    return df


# regex statement (will isolate all lines starting with course code)
regex_conditions = re.compile(r'([A-Z]{2}[0-9]{3})(.*)')  
line_split_conditions = '(#\d{3}\* |#\d{2}\* |\d{3}\* |\d{2}\* |\d{3} |\d{2} |#\d{3} |#\d{2} |# \+matric|#)'
df_2019 = data_handler_2019(regex, line_split_conditions, pdf_text)
df_2019.to_csv('data/cao_2019_final.csv')

print('There are {} lines matches'.format(len(df_2019[0])))

There are 930 lines matches


NB: it was verified as of 26/11/2021 that there were xxx courses exactly in the 930 CAO 2019 points list.

In [53]:
# Spot check a random row.
print(df_2019.iloc[50])

# Spot last row.
print(df_2019.iloc[-1])

0        CW708 
1    Law - LLB 
2          298 
3           328
Name: 50, dtype: object
0                                       WD230 
1    Mechanical and Manufacturing Engineering 
2                                         273 
3                                          348
Name: 929, dtype: object


0                                       WD230 
1    Mechanical and Manufacturing Engineering 
2                                         273 
3                                          348
Name: 929, dtype: object
